cql3/restrictions/statement_restrictions.cc

```diff
@@ -969,5 +975,229 @@ bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const que
     return (*_matcher)(data);
 }
 
+namespace wip {
+
+namespace {
+
+using cql3::selection::selection;
+
+/// Returns cdef's value from the fetched data.
+bytes_opt get_value(const column_definition* cdef, const selection& selection,
+                    const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+                    const std::vector<bytes_opt>& other_columns) {
+    switch (cdef->kind) {
+    case column_kind::partition_key:
+        return partition_key[cdef->id];
+    case column_kind::clustering_key:
+        return clustering_key[cdef->id];
+    case column_kind::static_column:
+    case column_kind::regular_column:
+        return other_columns[selection.index_of(*cdef)];
+    default:
+        throw exceptions::unsupported_operation_exception("Unknown column kind");
+    }
 }
+
+/// True iff columns' values equal t.
+bool equal(::shared_ptr<term> t, const idents& columns, const selection& selection,
+           const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+           const std::vector<bytes_opt>& other_columns, const query_options& options) {
+    if (auto multi = dynamic_pointer_cast<multi_item_terminal>(t)) {
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            if (rhs[i] != get_value(columns[i], selection, partition_key, clustering_key, other_columns)) {
+                return false;
+            }
+        }
+        return true;
+    } else  {
+        if (columns.size() != 1) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        return to_bytes_opt(t->bind_and_get(options)) ==
+                get_value(columns[0], selection, partition_key, clustering_key, other_columns);
+    }
 }
+
+/// True iff lhs is limited by rhs in the manner prescribed by op.
+bool limits(const bytes& lhs, const operator_type& op, const bytes& rhs, const abstract_type& type) {
+    if (!op.is_compare()) {
+        throw std::logic_error("limits() called on non-compare op");
+    }
+    const auto cmp = type.as_tri_comparator()(lhs, rhs);
+    if (cmp < 0) {
+        return op == operator_type::LT || op == operator_type::LTE || op == operator_type::NEQ;
+    } else if (cmp > 0) {
+        return op == operator_type::GT || op == operator_type::GTE || op == operator_type::NEQ;
+    } else {
+        return op == operator_type::LTE || op == operator_type::GTE || op == operator_type::EQ;
+    }
+}
```

Comment:
Breaking out op into the type system would have allowed us to pre-compute and store type.as_tri_comparator() and the expected result of cmp. So the run-time equivalent of this would be

    return _tricmp(lhs, rhs) == _expected_cmp;

Not suggesting to do it now, it's complicated enough already.